### PR TITLE
use DiagonalMatrix in MatrixFreeOperators

### PR DESCRIPTION
**Pull request by davydden**

### CODE REVIEW

File: include/deal.II/numerics/vector_tools.templates.h [hunk 1]
```diff
@@ -971,7 +971,7 @@ namespace VectorTools
       ReductionControl     control(rhs.size(), 0., 1e-12, false, false);
       SolverCG<LinearAlgebra::distributed::Vector<Number> > cg(control);
       typename PreconditionChebyshev<MatrixType, LocalVectorType>::AdditionalData data;
-      data.matrix_diagonal_inverse = mass_matrix.get_matrix_diagonal_inverse();
+      data.matrix_diagonal_inverse = mass_matrix.get_matrix_diagonal_inverse().get_vector();
```
Comment (davydden) [inline include/deal.II/numerics/vector_tools.templates.h]: @kronbichler If I am not mistaken, there is a better way of using `DiagonalMatrix` in `PreconditionChebyshev`.

File: include/deal.II/matrix_free/operators.h [hunk 2]
```diff
@@ -1028,7 +1025,7 @@ namespace MatrixFreeOperators
     Assert((Base<dim, Number>::data != NULL), ExcNotInitialized());
 
     unsigned int dummy = 0;
-    LinearAlgebra::distributed::Vector<Number> &inverse_diagonal_entries = Base<dim,Number>::inverse_diagonal_entries;
+    LinearAlgebra::distributed::Vector<Number> inverse_diagonal_entries;
```
Comment (kronbichler) [inline include/deal.II/matrix_free/operators.h]: You could still keep the variable `inverse_diagonal_entries` here if you do `this->inverse_diagonal_entries.get_vector()`. That way you avoid copying the vector and holding two vectors for a little while. (The same above for the mass matrix.)

File: include/deal.II/numerics/vector_tools.templates.h (hunk 1 above)
Comment (kronbichler) [inline include/deal.II/numerics/vector_tools.templates.h]: Yes, you would call 
```
data.preconditioner.reset(new DiagonalPreconditioner<LocalVectorType>());
*data.preconditioner = mass_matrix.get_diagonal_inverse();
```

If we had an `std_cxx11::shared_ptr<DiagonalPreconditioner<VectorType> >` in the `Base` class and returned the shared pointer, we could even share the storage between the two. Do you want to try that?

And one more remark: I recommend against using PreconditionChebyshev for solving a single system - at least in the mass matrix context. It needs to estimate eigenvalues by solving an auxiliary linear system. Until you have estimated the eigenvalues of the mass matrix, you have almost solved the system already. My suggestion is to use a Jacobi preconditioner, i.e., simply what you have already. I should have seen that last week already but it somehow slipped through... ;-)

File: include/deal.II/matrix_free/operators.h (hunk 2 above)
Comment (kronbichler) [inline include/deal.II/matrix_free/operators.h]: But there is a read-write variant available, see line 54 in the current github.

Comment (davydden) [inline include/deal.II/matrix_free/operators.h]: i must have missed that, will do, thanks!

**Comment by davydden** [comment]
this is ready for a second review, not a WIP anymore.

**Comment by kronbichler** [comment]
This looks good. I assume you do not want to make the `DiagonalMatrix<VectorType>` in the `MatrixFreeOperators::Base` class a shared pointer (at least for now)? I guess it's fine to postpone to another PR in case we find that useful. But we should do it not too far in the future, and definitely before the next release, because we should not change interfaces too often.

**Comment by davydden** [comment]
>  I assume you do not want to make the DiagonalMatrix<VectorType> in the MatrixFreeOperators::Base class a shared pointer (at least for now)

i would indeed postpone it. Feel free to open an issue and tag it for next release so that we don't forget about it.

**Comment by kronbichler** [comment]
OK, this is now #3343.

**Comment by davydden** [comment]
@kronbichler i broke a few tests, PR incoming...

**Pull request merged**